### ISSUE TITLE
FSDP does not move modules without parameters to device

### ISSUE THREAD

**Issue by awaelchli** [pytorch/pytorch#113113]
### 🐛 Describe the bug

I discovered that FSDP doesn't call `.to()` on submodules that have no parameters. This seems odd, since other modules get moved automatically even if they are not wrapped explicitly. 

The following example code demonstrates this:

```py
import os

import torch.cuda
import torch.nn as nn
import torch.distributed as dist
import torch.multiprocessing as mp
from torchmetrics import Accuracy
from torch.distributed.fsdp import FullyShardedDataParallel


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(10, 10)
        # This metric is an nn.Module, but doesn't have parameters
        self.metric = Accuracy(task="multiclass", num_classes=10)


def work(rank):
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "1234"
    dist.init_process_group("nccl", world_size=2, rank=rank)

    torch.cuda.set_device(rank)
    device = torch.device("cuda", rank)

    model = Model()
    model = FullyShardedDataParallel(mode, device_id=rank)

    assert not any(model.metric.parameters())

    assert model.linear.weight.device == device
    assert model.metric.tp.device == device  # fails!


def run():
    mp.spawn(work, nprocs=2)


if __name__ == "__main__":
    run()

```

I'd like to ask whether this is intentional for a good reason, and if not could it be changed?

### Versions

PyTorch version: 2.2.0.dev20230920+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.35

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-75-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 525.125.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          256
On-line CPU(s) list:             0-254
Off-line CPU(s) list:            255
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7742 64-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
Stepping:                        0
Frequency boost:                 enabled
CPU max MHz:                     2250.0000
CPU min MHz:                     0.0000
BogoMIPS:                        4499.83
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                  AMD-V
L1d cache:                       4 MiB (128 instances)
L1i cache:                       4 MiB (128 instances)
L2 cache:                        64 MiB (128 instances)
L3 cache:                        512 MiB (32 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-63,128-191
NUMA node1 CPU(s):               64-127,192-254
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] onnx==1.12.0
[pip3] onnxruntime==1.14.1
[pip3] pytorch-lightning==2.0.6
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20230920+cu121
[pip3] torchmetrics==1.0.3
[pip3] torchvision==0.17.0.dev20230920+cu121
[pip3] triton==2.0.0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-lightning         2.0.6                    pypi_0    pypi
[conda] pytorch-triton            2.1.0+6e4932cda8          pypi_0    pypi
[conda] torch                     2.2.0.dev20230920+cu121          pypi_0    pypi
[conda] torchmetrics              1.0.3                    pypi_0    pypi
[conda] torchvision               0.17.0.dev20230920+cu121          pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu

**Comment by awgu** [comment]
Is `model.metric.tp` a buffer?

I think it is probably an oversight. We should check if there are any parameters or buffers.

**Comment by awaelchli** [comment]
@awgu Apologies yes, of course the `device_id=rank` was missing. I must have messed with Ctrl+Z before submitting. 

You are right, `model.metric.tp` is not a buffer. And I asked the torchmetrics team and there is a good reason for it (they need to support metric states inside of lists, and buffers can only be tensors not lists). But this is an implementation detail of torchmetrics. 

I realize now that since these attributes in torchmetrics are just regular tensors, not buffers, we can't expect FSDP to move them automatically to the device. Torchmetrics moves them to the device when `.to()` is called on the metric object. There is of course a good reason why FSDP doesn't call `.to()` on its modules, because that would be recursive. So I understand now that this might not be something FSDP can fix. 

Perhaps we should just recommend that users don't put their torchmetrics as submodules of a FSDP model.

**Comment by awaelchli** [comment]
Perhaps FSDP could make the following check: If a submodule is not wrapped by FSDP and also does not have any descendant FSDP submodules, then it could be calling `.to()` on those submodules. What do you think?

**Comment by awgu** [comment]
@awaelchli Sorry for the delay. This is a good point! I did not think that users would override `nn.Module.to()`.

There is some nuance in this problem. FSDP currently has an `ignored_states` argument that would allow a user to partially ignore a module (e.g. some parameters ignored but others not). If we want to support that properly, FSDP in theory should not use `nn.Module.to()` to move parameters to GPU; rather, it should do it tensor by tensor.

For users overriding `.to()`, is it fair to ask them to add an extra `.cuda()` or `.to(...)` call after applying FSDP? Would that work in this case?

**Comment by awaelchli** [comment]
> There is some nuance in this problem. FSDP currently has an ignored_states argument that would allow a user to partially ignore a module (e.g. some parameters ignored but others not). If we want to support that properly, FSDP in theory should not use nn.Module.to() to move parameters to GPU; rather, it should do it tensor by tensor.

Yes, I agree :)

> For users overriding .to(), is it fair to ask them to add an extra .cuda() or .to(...) call after applying FSDP? Would that work in this case?

Yes I think that is fair. Currently the only use case where this came up is FSDP + torchmetrics + Lightning, and in Lightning I made a patch to do this `.to()` call for metric objects automatically.

**Comment by mori360** [comment]
Solved by answer https://github.com/pytorch/pytorch/issues/113113#issuecomment-1821624037
